com.unity.shadergraph/Editor/Drawing/PreviewManager.cs

```diff
@@ -243,276 +282,443 @@ public bool HandleGraphChanges()
             {
-                m_NodesToUpdate.Add(data.shaderData.node);
+                m_NodesNeedsRecompile.Add(data.shaderData.node);
             }
         }
 
-        void UpdateShaders()
+        private static readonly ProfilerMarker ProcessCompletedShaderCompilationsMarker = new ProfilerMarker("ProcessCompletedShaderCompilations");
+        void ProcessCompletedShaderCompilations()
         {
             // Check for shaders that finished compiling and set them to redraw
-            foreach (var renderData in m_RenderDatas.Values)
+            using (ProcessCompletedShaderCompilationsMarker.Auto())
+            using (var nodesCompiled = PooledHashSet<AbstractMaterialNode>.Get())
             {
-                if (renderData.shaderData.isCompiling)
+                foreach (var node in m_NodesCompiling)
                 {
-                    var isCompiled = true;
+                    PreviewRenderData renderData = GetPreviewRenderData(node);
+                    PreviewShaderData shaderData = renderData.shaderData;
+                    Assert.IsTrue(shaderData.passesCompiling > 0);
+
+                    if (shaderData.passesCompiling != renderData.shaderData.mat.passCount)
+                    {
+                        Debug.LogWarning("WARNING, shader pass count changed while compiling preview shader, preview may not complete");
+                    }
+
+                    // check that all passes have compiled
+                    var allPassesCompiled = true;
                     for (var i = 0; i < renderData.shaderData.mat.passCount; i++)
                     {
                         if (!ShaderUtil.IsPassCompiled(renderData.shaderData.mat, i))
                         {
-                            isCompiled = false;
+                            allPassesCompiled = false;
                             break;
                         }
                     }
 
-                    if (!isCompiled)
+                    if (!allPassesCompiled)
                     {
                         continue;
                     }
 
-                    // Force the material to re-generate all it's shader properties.
+                    // Force the material to re-generate all it's shader properties, by reassigning the shader
                     renderData.shaderData.mat.shader = renderData.shaderData.shader;
-
-                    renderData.shaderData.isCompiling = false;
+                    renderData.shaderData.passesCompiling = 0;
+                    renderData.shaderData.isOutOfDate = false;
                     CheckForErrors(renderData.shaderData);
-                    m_NodesToDraw.Add(renderData.shaderData.node);
+
+                    nodesCompiled.Add(renderData.shaderData.node);
 
                     var masterNode = renderData.shaderData.node as IMasterNode;
-                    masterNode?.ProcessPreviewMaterial(renderData.shaderData.mat);
+                    if (masterNode != null)
```

Comment:
You don't like the null-conditional operator? You could use pattern matching instead:
`if(rd.sd.node as IMasterNode masterNode)`

Reply:
Hmm, I must have added code in there and then removed it, I can change this back.